exporter/signalfx/signalfx.go

```diff
@@ -0,0 +1,303 @@
+	}
+	e.client.Sink.(*sfxclient.HTTPSink).AuthToken = e.opts.Token
+
+	if e.opts.ReportingDelay != 0 {
+		e.client.ReportingDelay(e.opts.ReportingDelay)
+	}
+
+	go e.client.Schedule(context.Background())
+
+	return e, nil
+}
+
+var _ view.Exporter = (*Exporter)(nil)
+
+// registerViews creates the view map and prevents duplicated views
+func (c *collector) registerViews(views ...*view.View) {
+	for _, thisView := range views {
+		sig := viewSignature(thisView.Name, thisView)
+		c.registeredViewsMu.Lock()
+		_, ok := c.registeredViews[sig]
+		c.registeredViewsMu.Unlock()
+		if !ok {
+			desc := sanitize(thisView.Name)
+			c.registeredViewsMu.Lock()
+			c.registeredViews[sig] = desc
+			c.registeredViewsMu.Unlock()
+		}
+	}
+}
+
+func (o *Options) onError(err error) {
+	if o.OnError != nil {
+		o.OnError(err)
+	} else {
+		log.Printf("Failed to export to SignalFx: %v", err)
+	}
+}
+
+// ExportView exports to the SignalFx if view data has one or more rows.
+// Each OpenCensus stats records will be converted to
+// corresponding SignalFx Metric
+func (e *Exporter) ExportView(vd *view.Data) {
+	if len(vd.Rows) == 0 {
+		return
+	}
+	e.c.addViewData(vd)
+
+	extractViewData(vd, e)
+}
+
+// toMetric receives the view data information and creates metrics that are adequate according to
+// graphite documentation.
+func (c *collector) toMetric(v *view.View, row *view.Row, vd *view.Data, e *Exporter) signalFxMetric {
+	switch data := row.Data.(type) {
+	case *view.CountData:
+		metric := signalFxMetric{
+			metricName:     sanitize(vd.View.Name),
+			metricType:     "cumulative_counter",
+			metricValueInt: data.Value,
+			timestamp:      time.Now(),
+			dimensions:     buildDimensions(row.Tags),
+		}
+		return metric
+	case *view.SumData:
+		metric := signalFxMetric{
+			metricName:     sanitize(vd.View.Name),
+			metricType:     "cumulative_counter",
+			metricValueInt: int64(data.Value),
+			timestamp:      time.Now(),
+			dimensions:     buildDimensions(row.Tags),
+		}
+		return metric
+	case *view.LastValueData:
+		metric := signalFxMetric{
+			metricName:  sanitize(vd.View.Name),
+			metricType:  "gauge",
+			metricValue: data.Value,
+			timestamp:   time.Now(),
+			dimensions:  buildDimensions(row.Tags),
+		}
+		return metric
+	default:
+		// TODO: add support for histograms (Aggregation.DistributionData).
+		e.opts.onError(fmt.Errorf("aggregation %T is not yet supported", data))
+		return signalFxMetric{}
+	}
+}
+
+// extractViewData extracts stats data
+func extractViewData(vd *view.Data, e *Exporter) {
+	for _, row := range vd.Rows {
+		metric := e.c.toMetric(vd.View, row, vd, e)
+		go sendRequest(e, metric)
+	}
+}
+
+// buildDimensions uses the tag values and keys to create
+// the dimensions used by SignalFx
+func buildDimensions(t []tag.Tag) map[string]string {
+	values := make(map[string]string)
+	for _, t := range t {
+		values[t.Key.Name()] = t.Value
+	}
+	return values
+}
+
+type collector struct {
```

Comment:
Instead of using this, would it be possible to use a Bundler like the Graphite and Stackdriver exporter? This has a well-defined behavior and doesn't require reimplementing the buffering behavior in every exporter.

Reply:
@ramonza Done